Google.Api.Gax.Rest/ClientBuilderBase.cs

```diff
@@ -0,0 +1,181 @@
+        {
+            ValidateAtMostOneNotNull("Only one source of credentials can be specified",
+                CredentialsPath, JsonCredentials, Credential);
+        }
+
+        /// <summary>
+        /// Validates that at most one of the given values is not null.
+        /// </summary>
+        /// <param name="message">The message if the condition is violated.</param>
+        /// <param name="values">The values to check for nullity.</param>
+        /// <exception cref="InvalidOperationException">More than one value is null.</exception>
+        protected void ValidateAtMostOneNotNull(string message, params object[] values)
+        {
+            int notNull = values.Count(v => v != null);
+            GaxPreconditions.CheckState(notNull < 2, message);
+        }
+
+        /// <summary>
+        /// Creates an initializer for the service. This method does not perform any validation.
+        /// </summary>
+        /// <returns>An initializer for the service.</returns>
+        protected virtual BaseClientService.Initializer CreateServiceInitializer() =>
+            new BaseClientService.Initializer
+            {
+                HttpClientInitializer = GetHttpClientInitializer(),
+                ApiKey = ApiKey,
+                ApplicationName = ApplicationName ?? GetDefaultApplicationName(),
+                BaseUri = BaseUri
+            };
+
+        /// <summary>
+        /// Creates an initializer for the service asynchronously. This method does not perform any validation.
+        /// </summary>
+        /// <returns>An initializer for the service.</returns>
+        protected virtual async Task<BaseClientService.Initializer> CreateServiceInitializerAsync(CancellationToken cancellationToken) =>
+            new BaseClientService.Initializer
+            {
+                HttpClientInitializer = await GetHttpClientInitializerAsync(cancellationToken).ConfigureAwait(false),
+                ApiKey = ApiKey,
+                ApplicationName = ApplicationName ?? GetDefaultApplicationName(),
+                BaseUri = BaseUri
+            };
+
+        /// <summary>
+        /// Obtains credentials synchronously. Override this method in a concrete builder type if more
+        /// credential mechanisms are supported.
+        /// </summary>
+        protected virtual IConfigurableHttpClientInitializer GetHttpClientInitializer()
+        {
+            if (Credential != null)
+            {
+                return Credential;
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) :
+                JsonCredentials != null ? GoogleCredential.FromJson(JsonCredentials) :
+                null; // Use default credentials (maybe - see below)
+            // While we accept any credentials that are specified even if there's an API key,
+            // we don't try to load default credentials.
+            if (ApiKey != null && unscoped is null)
```

Comment:
Really need an `isnt` operator for consistency.

Reply:
Ack. `ApiKey is object` works, but it's a bit odd-looking to start with.